driftbase/api/friendships.py

```diff
@@ -165,35 +164,43 @@ def delete(self, friendship_id):
 
 @bp.route('/invites', endpoint='invites')
 class FriendInvitesAPI(MethodView):
+    class CreateInviteSchema(ma.Schema):
+        player_id = ma.fields.Integer(required=False, metadata=dict(description="The receiving player of the invite. Optional."))
+        expiration_time_seconds = ma.fields.Integer(required=False, metadata=dict(description="The expiration time of the invite in seconds."))
+        token_format = ma.fields.String(required=False, metadata=dict(description="The format of the token. Supported values: 'uuid' and 'wordlist'."))
+        worldlist_number_of_words = ma.fields.Integer(required=False, metadata=dict(description="How many words the token should be if the token format is 'wordlist'"))
 
     @bp.response(http_client.OK, InviteSchema(many=True))
     def get(self):
         """ List invites sent by current player """
         CorePlayer2 = aliased(CorePlayer)
         return g.db.query(FriendInvite, CorePlayer.player_name, CorePlayer2.player_name). \
-            join(CorePlayer, CorePlayer.player_id == FriendInvite.issued_to_player_id). \
+            join(CorePlayer, CorePlayer.player_id == FriendInvite.issued_to_player_id, isouter=True). \
             join(CorePlayer2, CorePlayer2.player_id == FriendInvite.issued_by_player_id). \
             filter(FriendInvite.issued_by_player_id == int(current_user["player_id"]),
                    FriendInvite.expiry_date > datetime.datetime.utcnow(),
                    FriendInvite.deleted.is_(False))
 
-    def post(self):
+    @bp.arguments(CreateInviteSchema)
+    def post(self, args):
         """
         New Friend token
         """
         sending_player_id = int(current_user["player_id"])
         try:
-            receiving_player_id = int(request.args.get("player_id"))
+            receiving_player_id = int(args.get("player_id"))
             self._validate_friend_request(receiving_player_id)
         except TypeError:
             receiving_player_id = None
 
-        token = str(uuid.uuid4())
-        expires_seconds = DEFAULT_INVITE_EXPIRATION_TIME_SECONDS
-        config = g.conf.tenant.get('friends')
-        if config:
-            expires_seconds = config['invite_expiration_seconds']
-        expires_seconds = expires_seconds
+        token_format = args.get("token_format") or _get_tenant_config_value("invite_token_format")
+        if token_format == "uuid":
+            token = str(uuid.uuid4())
+        elif token_format == "wordlist":
+            number_of_words = args.get("worldlist_number_of_words") or _get_tenant_config_value("invite_token_worldlist_number_of_words")
```

Comment:
I think it's sensible to a) restrict this to min 3 words to avoid silly 1 word codes, b) make sure the token is unique and retry on failure, as that's the practical effect of using the uuid earlier, it just wasn't enforced.

Reply:
I think 2 word combos are valid in some cases, but yes 1 word makes no sense.
Will add some sort of retry logic for the wordlist.

Reply:
Will need to add a unique constraint to the schema too, with an alembic up/downgrade script. I think Einsi knows how.

Reply:
Then I think we should actually delete the invites when they are deleted and if the token is in use, but expired, then I think we should delete the expired token and use the conflicting token value.

Reply:
That just gives you two cases, when you have a collision, retry if it's not expired, or generate a new one if it is. If you don't ever re-use, then there's a single path for all cases, always retry

Reply:
Min number of words validation still needed here